python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py

```diff
@@ -0,0 +1,536 @@
+                    timestamp_ns=time.time_ns(),
+                    count=count,
+                    request_id=request_id,
+                    exception=e,
+                )
+            )
+
+    def _add_terminate_errors(
+        self,
+        ids: List[CloudInstanceId],
+        request_id: str,
+        details: str,
+        e: Optional[Exception] = None,
+    ) -> None:
+        """
+        Adds terminate errors to the error queue.
+
+        Args:
+            ids: The ids of the nodes that failed to terminate.
+            request_id: The request id of the terminate request.
+            details: The details of the error.
+            e: The exception that caused the error.
+        """
+        for id in ids:
+            self._terminate_errors_queue.append(
+                TerminateNodeError(
+                    cloud_instance_id=id,
+                    details=details,
+                    timestamp_ns=time.time_ns(),
+                    request_id=request_id,
+                    exception=e,
+                )
+            )
+
+    def _sync_with_api_server(self) -> None:
+        """Fetches the RayCluster resource from the Kuberay API server."""
+        self._ray_cluster = self._get(f"rayclusters/{self._cluster_name}")
+        self._cached_instances = self._fetch_instances()
+
+    @property
+    def ray_cluster(self) -> Dict[str, Any]:
+        return copy.deepcopy(self._ray_cluster)
+
+    @property
+    def instances(self) -> Dict[CloudInstanceId, CloudInstance]:
+        return copy.deepcopy(self._cached_instances)
+
+    @staticmethod
+    def _get_workers_groups_with_deletes(
+        ray_cluster_spec: Dict[str, Any], node_set: Set[CloudInstanceId]
+    ) -> Tuple[Set[NodeType], Set[NodeType]]:
+        """
+        Gets the worker groups that have pending deletes and the worker groups that
+        have finished deletes.
+
+        Returns:
+            worker_groups_with_pending_deletes: The worker groups that have pending
+                deletes.
+            worker_groups_with_finished_deletes: The worker groups that have finished
+                deletes.
+        """
+
+        worker_groups_with_pending_deletes = set()
+        worker_groups_with_finished_deletes = set()
+
+        worker_groups = ray_cluster_spec["spec"].get("workerGroupSpecs", [])
+        for worker_group in worker_groups:
+            workersToDelete = worker_group.get("scaleStrategy", {}).get(
+                "workersToDelete", []
+            )
+            if not workersToDelete:
+                # No workers to delete in this group.
+                continue
+
+            node_type = worker_group["groupName"]
+            for worker in workersToDelete:
+                if worker in node_set:
+                    worker_groups_with_pending_deletes.add(node_type)
+                    break
+
+            # All to delete in the worker spec have been deleted.
+            worker_groups_with_finished_deletes.add(node_type)
```

Comment:
This is weird. All worker groups belonging to `worker_groups_with_pending_deletes` will also be added to `worker_groups_with_finished_deletes`. These two sets should be mutually exclusive.

Reply:
oh good catch! i wonder how come it wasn't caught with my test.

Reply:
hmm, i think this doesn't affect correctness since if any worker groups with pending deletes, we will not scale. So when there's no worker groups with pending deletes, all the worker groups will have finished deletes.

But yeah, this could be made mutually exclusive.